pubsublite/streams.go

```diff
@@ -0,0 +1,333 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package pubsublite
+
+import (
+	"context"
+	"io"
+	"reflect"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	gax "github.com/googleapis/gax-go/v2"
+)
+
+// streamStatus is the state of a gRPC client stream. A stream starts off
+// uninitialized. While it is active, it can transition between reconnecting and
+// connected due to retryable errors. When a permanent error occurs, the stream
+// is terminated and cannot be reconnected.
+type streamStatus int
+
+const (
+	streamUninitialized streamStatus = 0
+	streamReconnecting  streamStatus = 1
+	streamConnected     streamStatus = 2
+	streamTerminated    streamStatus = 3
+)
+
+// streamHandler provides hooks for different streaming RPCs (e.g. publish,
+// subscribe, streaming cursor, etc) to use retryableStream. All Pub/Sub Lite
+// streaming RPCs implement a similar handshaking protocol, where an initial
+// request and response must be transmitted before other requests can be sent
+// over the stream.
+//
+// streamHandler methods must not be called while holding the retryableStream.mu
+// in order to prevent the streamHandler calling back into the retryableStream
+// and deadlocking.
+type streamHandler interface {
+	// newStream implementations must create the client stream with the given
+	// (cancellable) context.
+	newStream(context.Context) (grpc.ClientStream, error)
+	initialRequest() interface{}
+	validateInitialResponse(interface{}) error
+
+	// onStreamStatusChange is used to notify stream handlers when the stream has
+	// changed state. In particular, the `streamTerminated` state must be handled.
+	// retryableStream.Error() returns the error that caused the stream to
+	// terminate. Stream handlers should perform any necessary reset of state upon
+	// `streamConnected`.
+	onStreamStatusChange(streamStatus)
+	// onResponse forwards a response received on the stream to the stream
+	// handler.
+	onResponse(interface{})
+}
+
+// retryableStream is a wrapper around a bidirectional gRPC client stream to
+// handle automatic reconnection when the stream breaks.
+//
+// A retryableStream cycles between the following goroutines:
+//   Start() --> reconnect() <--> listen()
+// terminate() can be called at any time, either by the client to force stream
+// closure, or as a result of an unretryable error.
+type retryableStream struct {
+	// Immutable after creation.
+	ctx          context.Context
+	handler      streamHandler
+	responseType reflect.Type
+	timeout      time.Duration
```

Comment:
Why is a timeout and a context needed?

Reply:
The timeout is passed to streamRetryer, to ensure that we don't loop retrying forever. The user has control over the timeout, it will basically emulate [pubsub.PublishSettings.Timeout](https://godoc.org/cloud.google.com/go/pubsub#PublishSettings).

The context is used to create the cancellable context, for closing the stream client-side, and also to pass to gax.Sleep() during retries.